### ISSUE TITLE
Attempting to add user to group or just list groups is resulting in a 403.

### ISSUE THREAD

**Issue by silverl** [squid808/gShell#4]
I'm trying to add a user to a group. This failed with a 403. I then tried getting the list of groups and met with the same 403.

I previously used Invoke-ScopeManager to request all permissions in admin:directory_v1, and that went fine.

I have successfully used gShell to set a user's password, so at least I've proven something works.

Here's what happens when I attempt to add a user to a group:
```
PS C:\> Add-GAGroupMember -GroupName mygroup@mydomain.com -Domain mydomain.com -username jsmith -Role MEMBER
Add-GAGroupMember : Google.Apis.Requests.RequestError
Insufficient Permission [403]
Errors [
    Message[Insufficient Permission] Location[ - ] Reason[insufficientPermissions] Domain[global]
]
At line:1 char:1
+ Add-GAGroupMember -GroupName mygroup@mydomain.com -Domain mydomain.com -u ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Add-GAGroupMember], GoogleApiException
    + FullyQualifiedErrorId : Google.GoogleApiException,gShell.Cmdlets.Directory.GAGroupMember.AddGAGroupMember

**Comment by squid808** [comment]
I'm unable to reproduce - using the same exact line you provided (with my own user and domain swapped in) it completed successfully.

First, double check what scopes you have authenticated with the following (this assumes you only have the one user authenticated with gShell so far):

Make sure the account you authenticated gShell with has the proper permissions in Google Apps to add group members (are you an admin?)

If you are able to do this normally through the GUI interface try removing the gShell_OAuth2.bin file in *%userprofile%\AppData\Local\gShell* and re-authenticating yourself, maybe something went wrong with Invoke-ScopeManager.

If still no luck, try a different user from Google who has the proper admin rights to see if that helps.

**Comment by silverl** [comment]
I'm the domain admin. I've only started using gShell today, and have only authenticated using my account.

I ran your Get-gShellUser command. There are the scopes I'm authenticated with.

```
https://www.googleapis.com/auth/admin.directory.device.chromeos
https://www.googleapis.com/auth/admin.directory.device.chromeos.readonly
https://www.googleapis.com/auth/admin.directory.device.mobile
https://www.googleapis.com/auth/admin.directory.device.mobile.action
https://www.googleapis.com/auth/admin.directory.device.mobile.readonly
https://www.googleapis.com/auth/admin.directory.group
https://www.googleapis.com/auth/admin.directory.group.member
https://www.googleapis.com/auth/admin.directory.group.member.readonly
https://www.googleapis.com/auth/admin.directory.group.readonly
https://www.googleapis.com/auth/admin.directory.notifications
https://www.googleapis.com/auth/admin.directory.orgunit
https://www.googleapis.com/auth/admin.directory.orgunit.readonly
https://www.googleapis.com/auth/admin.directory.user
https://www.googleapis.com/auth/admin.directory.user.alias
https://www.googleapis.com/auth/admin.directory.user.alias.readonly
https://www.googleapis.com/auth/admin.directory.user.readonly
https://www.googleapis.com/auth/admin.directory.user.security
https://www.googleapis.com/auth/admin.directory.userschema
https://www.googleapis.com/auth/admin.directory.userschema.readonly
https://www.googleapis.com/auth/userinfo.email
```
I've deleted the .bin fine and re-authed. I still get the 403.

**Comment by squid808** [comment]
That all looks good from what I can see. The only other thing I can think of at this time is to go to [The API Documentation Page](https://developers.google.com/admin-sdk/directory/v1/reference/members/insert) and use the *Try It!* section at the bottom. You'd use it in a very similar way as you would gShell - click the slider to authenticate yourself (it will automatically choose the two scopes you need) and then put in the information to complete the request. Let me know if that works - I apologize I'm not able to give you a more solid idea of what is happening, so I'm still trying to narrow things down.

**Comment by squid808** [comment]
Just wanted to check in on this. I've heard from another user that they were having a similar issue, but after re-authenticating it worked without an issue. If you're still having problems, can you 1) try uninstalling your current version, removing the .bin file and [installing version 0.0.6](https://github.com/squid808/gShell/wiki/Downloads) to see if this might be a version dependent issue, and 2) [check out this issue on Stack Overflow](http://stackoverflow.com/questions/16970917/why-is-google-calendar-api-oauth2-responding-with-insufficient-permission) and see if that solution is of any help.

**Comment by squid808** [comment]
It's been a while with no response, please give [Version 0.0.8.0](https://github.com/squid808/gShell/wiki/News) a try - it fixes a lot of previous auth issues. Also take a look at [Getting Started](https://github.com/squid808/gShell/wiki/Getting-Started) to make sure you've covered all steps if you're still getting a 403. Otherwise, let me know and I'll happily do a remote session to see if we can't figure it out.

**Comment by Allieway** [comment]
I am actually having this exact same issue now. It persists after reinstalling 0.9.5, 0.8.1, removing the bin, and attempting to try the same thing on the Documentation API works perfectly(as well as multiple other write operations working on the API site) . I followed the issue on Stack Overflow by revoking access and adding back. 

The problem seems to be that once I give access to the read-only scopes Gshell no longer opens up oauth 2 authentication to access the red-write scopes, instead it just says `admin:directory_v1 has been authenticated and saved.` If I reinstall gshell and attempt to authenticate the read-write scopes first before the read only it opens the browser, says `Received verification code. You may now close this window.` but then gives me the insufficient permission error regardless of what I call(whether it be invoke-scopemanager or Another command to invoke the scope manager automatically.

I am super admin in the google apps domain, on windows 10 64bit, running powershell 5, with 0.9.5 gshell

**Comment by squid808** [comment]
Hi @Allieway - you mentioned you're having the issue _now_ which makes it sound like you had things working prior to this - is that the case? If so, are you able to do other read/write operations in gShell at all?

**Comment by Allieway** [comment]
Sorry I meant now as in compared to over a year ago when it was originally reported. I have not been able to get it working yet, and no, I cannot seem to do other read write operations.

**Comment by squid808** [comment]
Thanks for the update. I'm thinking the issue may be in how the browser is trying to relay the information back to PowerShell. This is all managed in some of the underlying Google code, but my understanding is that when we try to authenticate it sets up some small listening server and once the browser is done it redirects information back to that listener... thing. (I apologize, but never had to get too in-detail for it!)

So, with that in mind, my troubleshooting process is:

- Check to see if any information is making it back to gShell and being saved
- If not, troubleshoot the browser-powershell connection
- If so, troubleshoot the information that's coming back

If you check _%userprofile%\Documents\WindowsPowerShell\Modules\gShell\gShell_OAuth2.json_ (which is created by default when you first authenticate) you should see a json structure containing the information saved for your domain. Can you take a look in this file and see if you see information? The structure is:
domains > [your domain] > users > [your user] > tokenAndScopes > [the api in question]

Under the API level you should see _scopes_, _tokenString_ and _token_. The _scopes_ likely has a list of things that you selected, do the tokenString and token both exist and have data in them?

**Comment by Allieway** [comment]
Both do exist and have data in them, but only include the readonly scopes that went correctly with authentication. 

When I reinstall gshell and try to start with either both read and read/write scopes or just read/write scopes the OAuth2.json file only has my client_id and client_secret. No scopes, token string, or token. It looks like it didn't even populate the domains or users, so that would be that information is not being made back to gshell and being saved(when I select read/write scopes)

**Comment by squid808** [comment]
Ok, thank you. What do you mean about the scopes that went correctly with authentication? Are you authenticating twice, once with read-only and a second time with read-write on the same API? If that's the case, it would only store the most recent - the scopes get re-written each time you do it. If you want to include a combination of read-write and read-only then have the wizard show the full list and choose them individually from there. Otherwise if you're doing all of read-write then you have no need for the read-only scopes, since the read-write scopes cover that.

That said, let's try to get you running with a basic Get-GAUser command for now, if you're willing. I admit I'm flying blind for your issue as I can't reproduce or troubleshoot it on my end. (For the duration, you may want to set up a function like Set-MySecrets that would handle setting up the secrets for you, in case you end up doing this repeatedly - it can get annoying otherwise).

0) Make sure you've done everything necessary in the [Getting-Started](https://github.com/squid808/gShell/wiki/Getting-Started) guide including:

- You have enabled API access for your domain
- In the Google Developer's Console project, you have enabled the Directory API for use and have used the credentials for that project

1) You mentioned the _Documentation API_ worked on the google site - did you meant the Directory API? Either way, let's focus on getting a single user. Using [this page](https://developers.google.com/admin-sdk/directory/v1/reference/users/get), can you return one single user with the Try It Now option?

2) Close PowerShell, completely uninstall gShell, delete the folders that it left behind _(%userprofile%\Documents\WindowsPowerShell\Modules\gShell_), reinstall the [newest gShell (0.9.5)](https://github.com/squid808/gShell/wiki/Downloads).

3) Launch PowerShell, import the module, and set your client secrets. 

4) After setting your client secrets, try a basic cmdlet: `Get-GAUser _someUser_`

5) Run through the scope wizard, choose the option _w_ to get all read-write scopes, and then choose 0 to select them all. Hit enter and wait for the browser to open. **What browser is opening?**

6) The list of scopes should be listed here. Make sure that it's for the domain account, and switch to that account if necessary. _This should be the same account that you verified is working in step 1._

7) Once you're set up with the proper account, click the button for 'yes' or 'I approve' or whatever the affirmative button is. It should give the information about being able to close the browser window as you mentioned.

8) Go back to PowerShell - confirm that you're seeing the 403 error here?

If that's the case, try closing PowerShell, deleting the _gSHell_OAuth2.json_ file, and trying steps 3 through 8 again. But this time, when your default browser pops open, copy the link address it gives you, go to another browser and try it in that browser?

If _that_ doesn't work, can you make a private Gist or something and share a sanitized version of your _gSHell_OAuth2.json_ file with me?

**Comment by Allieway** [comment]
I only try to authenticate multiple times due to it not working when I try authenticating with all, read/write, or read-only scopes. `Would you like to view all 27 scopes [a], the 12 read-only scopes [r] or 15 read-write scopes [w]?` Pressing a results in a 403 error, pressing r results in it working, but only being able to use read-only scopes.

0) I have done everything and triple checked

1) Yes I do you mean Directory API, yes I can return the user.

After quite a bit of testing(thanks for the recommendation of the secret function!) it seems that the issue is when I attempt to press a instead of w, I guess I had not attempted to press w yet as I assumed all scopes were better than only read-write(I didn't realize read-write was inclusive of read only). So using only read-write or only read works perfectly. It is when pressing all 403 is returned. Is this because having the read-only scopes overrides the read-write scopes? If that is the case why is all even an option as it seems like it does nothing different than read-only?

**Comment by Allieway** [comment]
This happened again today and I had to follow both your uninstall instructions and remove the app from my account as recommended on the stack overflow link your posted. it happened when I attempted to add gmail with Invoke-ScopeManager. I started getting 403 on any of the things I had authenticated before hand. I had to get some work done so I wasn't able to test further as I wanted to fix it quickly.

**Comment by squid808** [comment]
Thanks for your patience! To touch on some of the points you raised:

> It is when pressing all 403 is returned. Is this because having the read-only scopes overrides the read-write scopes?

From my testing, it's not that (I authenticated for just the user and user.readonly and there was no conflict issue). It has something to do with selecting all 27 of the scopes available. I tried that and sure enough, I got the 403 error. This is good!

> If that is the case why is all even an option as it seems like it does nothing different than read-only?

This is to allow people to pick and choose what gShell has access to, if they wanted to customize it for a lower-tier support, for instance. You could, in theory, set it up so someone has the ability to read everything, but create/modify user aliases.

> it happened when I attempted to add gmail with Invoke-ScopeManager.

Yeah, that may be my bad. It seems that, for whatever reason, once you've authenticated something for a domain, Invoke-ScopeManager doesn't actually do anything to overwrite your scopes. I've created a [new issue for ScopeManager here](https://github.com/squid808/gShell/issues/19) that may be affecting you - sorry about that! I'll be looking in to that as well.

I'll let you know when I have some updates!

**Comment by squid808** [comment]
Good news!

First, I fixed Issue #19 where ScopeManager wasn't working as intended which will come out in 0.9.7.0 when I fix a few other things I'll bundle in with it, hopefully soon. That's piece 1 of the puzzle.

Piece 2 - I had a bit of time to look in to this tonight, and it seems I've found the issue here. In your specific scenario, when you are choosing all 27 scopes, we're actually trying to get 28 - we always add in the scope for the [Oauth2 API's userinfo.email scope](https://developers.google.com/identity/protocols/googlescopes#oauth2v2) so that in certain conditions if you don't supply a domain or username and nothing is saved, we can still figure it out.

However, there is a limit to how many scopes a single application can request access to - roughly 25 or 26. That limit is what caused me to set up ScopeManager the way it is, where each API has to be authenticated separately. At the time, that was fine because all of the APIs were under that limit. But then they added the calendar API calls to the Directory API, pushing it up to a total of 27.

Now, we hit that limit and it silently drops the last couple of scopes on the list. The very last one on the list is the userinfo.email scope, aaaaaaaand that's what is failing and giving you the 403 error.

So, the good news is, things are 'working as intended' on both my end and Google's end. For the unlucky, I've added in a warning that you have to specifically say yes to, in order to continue warning about the large number of scopes.

At this point, we can proceed with the following:

## Solution
Make sure that for any one given API, you are not authenticating for more scopes than you absolutely need to - try to stick with a max of just read or read-write. Duplicate scopes may cause you to go over a scope threshhold outside the handling ability of gShell, and Google may drop necessary scopes! In cases where you have over 20 scopes selected, a warning will show up and advise you to change your choices.

**Comment by Allieway** [comment]
You're the best :D Thanks so much!
